### PR TITLE
Streaming and on_llm_new_token calling for LlamaCpp used for Alpaca

### PR DESCRIPTION
**Pull request by lukestanley**
## Summary

This adds streaming support to the LlamaCpp LLM class. It also improves the LlamaCpp class with a new `_get_parameters` method and updated tests to cover streaming tokens and the invocation of the `on_llm_new_token` callback.

### tests/integration_tests/llms/test_llamacpp.py

- Added `test_llamacpp_streaming` to test streaming tokens from LlamaCpp
- Added `test_llamacpp_streaming_callback` to test that streaming correctly invokes the `on_llm_new_token` callback

I'm new so if I've missed something, please kindly point it out.

### CODE REVIEW

**Comment by lukestanley** [comment]
I am fixing some linting errors on this.

File: langchain/llms/llamacpp.py [hunk 1]
```diff
@@ -195,3 +213,99 @@ def _call(self, prompt: str, stop: Optional[List[str]] = None) -> str:
...
+        Yields:
+            An async iterable of generated token strings.
+        """
+        params = self._prepare_params(stop, **config)
+
+        if stream:
+            for generation in self.client(prompt=prompts[0], stream=stream, **params):
+                chunk = generation['choices'][0]['text']
+            if self.verbose:
+                print(chunk, end="", flush=True)
```
Comment (vowelparrot) [inline langchain/llms/llamacpp.py:241]: Please remove the `print` statements

File: langchain/llms/llamacpp.py [hunk 2]
```diff
@@ -195,3 +213,99 @@ def _call(self, prompt: str, stop: Optional[List[str]] = None) -> str:
...
+            yield chunk
```
Comment (vowelparrot) [inline langchain/llms/llamacpp.py:242]: This `yield` statement ought to be indented once more, right?

File: tests/integration_tests/llms/test_llamacpp.py [hunk 3]
```diff
@@ -32,3 +35,32 @@ def test_llamacpp_inference() -> None:
...
+def test_llamacpp_streaming() -> None:
+    """Test streaming tokens from LlamaCpp."""
+    model_path = get_model()
+    llm = LlamaCpp(model_path=model_path, max_tokens=10)
+    generator = llm.stream("Q: How do you say 'hello' in German? A:'",stop=["'"])
+
+    assert isinstance(generator, Generator)
+
+    for token in generator:
+        assert isinstance(token, str)
```
Comment (vowelparrot) [inline tests/integration_tests/llms/test_llamacpp.py:48]: We should collect the tokens and assert the result is non-empty

**Review by vowelparrot — CHANGES_REQUESTED** [review]
Could you add a streaming examplesto the notebook as well? Thank you for the PR!

**Comment by lukestanley** [comment]
Thank you for looking, I think I made a mistake when fixing the merge conflict! I will fix mypy issues and improve the tests. I'm learning more about the way the LLM classes are meant to work! I am converting this back to draft!

**Comment by lukestanley** [comment]
> Could you add a streaming examplesto the notebook as well? Thank you for the PR!

I can try doing this, also I think a gif could be helpful with some chain being used.

**Comment by lukestanley** [comment]
Thanks for earlier reviews @vowelparrot  this should run much more happily now. I've simplified it a fair bit by throwing out the async stuff for now and just basing it around the stream generator. It also is more similar to the OpenAI LLM class now. I've implemented length checks as suggested too. Linting is passing for me as is the integration test.

File: langchain/llms/llamacpp.py [hunk 4]
```diff
@@ -167,31 +195,35 @@ def _call(self, prompt: str, stop: Optional[List[str]] = None) -> str:
...
         """
-
-        params = self._default_params
-        if self.stop and stop is not None:
-            raise ValueError("`stop` found in both the input and default params.")
-        elif self.stop:
-            params["stop_sequences"] = self.stop
-        elif stop:
-            params["stop_sequences"] = stop
+        print("Prompt: ", prompt)
```
Comment (vowelparrot) [inline langchain/llms/llamacpp.py:202]: Will want to remove print statements before landing

File: langchain/llms/llamacpp.py [hunk 5]
```diff
@@ -154,6 +157,31 @@ def _llm_type(self) -> str:
...
+            Dictionary containing the combined parameters.
+        """
+
+        # Raise error if stop sequences are in both input and default params
+        if self.stop and stop is not None:
+            raise ValueError("`stop` found in both the input and default params.")
+        
+        params = self._default_params
+        
+        # llamaa_cpp expects the "stop" key not this, so we remove it:
```
Comment (vowelparrot) [inline langchain/llms/llamacpp.py:177]:
```suggestion
         # llama_cpp expects the "stop" key not this, so we remove it:
```

**Review by vowelparrot — COMMENTED** [review]
Streaming is working locally - well done! Will have to clean up some of the linting errors but looks mostly there

File: langchain/llms/llamacpp.py (hunk 4 above)
Comment (lukestanley) [reply langchain/llms/llamacpp.py]: Oops, I was testing with load_summarize_chain and forgot this was in! Fixed now.

**Comment by lukestanley** [comment]
> Could you add a streaming examplesto the notebook as well? Thank you for the PR!

I improved the docstring for the stream method, which is not yet demoed but is very similar except it can include the logprobs. 

I added the callback demo. If people play it themselves they should see something like this (starts at 14 seconds):

https://user-images.githubusercontent.com/306671/233401708-be27b8a3-08d0-4361-a198-d8aa381e38fd.mp4

@vowelparrot

**Review by vowelparrot — APPROVED** [review]
Beautiful. Thank you for your patience!

**Pull request merged**